triggers/daml/ClearingTrigger.daml

```diff
@@ -146,3 +162,52 @@ collectDeposits restFunds (used, (nextDeposit::remaining))
                                 ((nextDeposit :: used), remaining)
   where
     nextAmount  = (.asset.quantity) . snd $ nextDeposit
+
+
+-- |Calculates the FairValue for all Derivatives known to Exchange
+handleCalculationRequests : Party -> ContractPair Clearing.FairValueCalculationRequest -> TriggerA ClearingState ()
+handleCalculationRequests party (requestCid, rq@Clearing.FairValueCalculationRequest{..}) = do
+  debug "handling calculation requests..."
+  allListings <- query @Listing.T
+  let optListingIdSet = Set.fromList <$> optListingIds
+  let listings = case optListingIdSet of
+        (Some ids) -> filterField (.listingId) (`Set.member` ids) allListings
+        None       -> allListings
+
+  forA_ listings (handleFairValueCalculation party currency upTo)
+  emitExerciseCmd requestCid Clearing.FairValueCalculationRequest_Ack
+  return ()
+
+-- |Calculates a FairValue for a single derivative.
+-- Order of precedence:
+--    1. The price of the most recent trade before the specified 'upTo'
+--    2. If there are no trades, the most recent 'FairValue'
+--    3. If there are no 'FairValue', request manual calculation
+--
+-- TODO: Allow user to specify a 'from' Time interval before which trades and
+-- fairvalues will be ignored.
+handleFairValueCalculation : Party -> Id -> Time -> ContractPair Listing.T -> TriggerA ClearingState ()
+handleFairValueCalculation party currency upTo (marketPairCid,l@(Listing.Listing{..})) = do
+  debug ("calculating FVR: " <> l.listingId)
+
+  fairValues <- sortOn ((.upTo) . snd)
+            <$> filterQuery @Clearing.FairValue (.listingId) (== l.listingId)
+
+  trades     <- sortOn (Down . (.timeNovated) . snd)
+            <$> filterQuery @Clearing.ClearedTradeSide (.order.details.symbol) (== l.listingId)
+
+  let relevantTrades = filter (\(_,ct) -> (subTime ct.timeNovated upTo) < (seconds 0)) trades
+  timestamp <- getTime
+
+  case getBestPrice relevantTrades fairValues of
+    (Some bp) -> dedupCreate $ Clearing.FairValue with provider = party, price = bp; ..
+    None      -> dedupCreate $ Clearing.ManualFairValueCalculation with provider = party,  ..
+
+  return ()
+
+  where
+    getBestPrice : [ContractPair Clearing.ClearedTradeSide] -> [ContractPair Clearing.FairValue] -> Optional Decimal
+    getBestPrice ts fvs
+      | not (null ts)  = Some . (.price) . (.execution) . snd . head $ ts
```

Comment:
```suggestion
      | not (null ts)  = Some . price . execution . snd . head $ ts
```
Does that not work?

Reply:
It does not...`(.price)` is a field accessor, I believe it's the only way of doing `contract.field` as a single function, at least for a `template`.

Reply:
Actually, I think you _can_ use the accessors as functions without the `.`, but since we aren't bothering making field names unique, it won't infer which function to use...so since multiple templates in the `Clearing` namespace have a `price` field, you get an ambiguous occurrence error.